runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -181,6 +199,111 @@ JITServerAOTDeserializer::deserialize(SerializedAOTMethod *method, const std::ve
+         {
+         // Add this runtime-generated ROMClass to the map so that it can be found when deserializing AOT methods
+         OMR::CriticalSection cs(_generatedClassesMonitor);
+
+         // Find the entry for this class loader and deterministic class name prefix
+         auto it = _generatedClasses.find({ ramClass->classLoader, StringKey(nameStr, prefixLength) });
+         if (it == _generatedClasses.end())
+            {
+            // Allocate a persistent copy of the name prefix for the new GeneratedClassMap instance
+            auto namePrefix = (uint8_t *)jitPersistentAlloc(prefixLength);
+            if (!namePrefix)
+               throw std::bad_alloc();
+            memcpy(namePrefix, nameStr, prefixLength);
+
+            try
+               {
+               // Add a new GeneratedClassMap instance for this class loader and name prefix
+               it = _generatedClasses.emplace_hint(it, std::piecewise_construct,
+                  std::forward_as_tuple(ramClass->classLoader, StringKey(namePrefix, prefixLength)),
+                  std::forward_as_tuple(namePrefix)
+               );
+               }
+            catch (...)
+               {
+               jitPersistentFree(namePrefix);
+               throw;
+               }
+            }
+
+         // Compute the deterministic ROMClass hash for this class
+         JITServerROMClassHash hash(ramClass->romClass, classLoadTRMemory(), TR_J9VMBase::get(_jitConfig, vmThread), true);
+         // Add a new entry to the GeneratedClassMap for this RAMClass
+         auto h_r = it->second._classHashMap.insert({ hash, ramClass });
+         if (h_r.second)
+            {
+            try
+               {
+               // Establish two-way mapping from the RAMClass so that it can be invalidated at class unload
+               auto p_r = it->second._classPtrMap.insert({ ramClass, hash });
+               TR_ASSERT_FATAL(p_r.second, "Duplicate generated class %p", ramClass);
+               }
+            catch (...)
+               {
+               it->second._classHashMap.erase(h_r.first);
+               throw;
+               }
+            }
+         else
+            {
+            TR_ASSERT_FATAL(ramClass->romClass == h_r.first->second->romClass,
+                            "Duplicate generated class %p != %p ROMClass mismatch: %p != %p",
+                            ramClass, h_r.first->second, ramClass->romClass, h_r.first->second->romClass);
+
+            if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+               {
+               char buffer[ROMCLASS_HASH_BYTES * 2 + 1];
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Duplicate generated class %.*s %p hash %s",
+                                              nameLen, nameStr, ramClass, hash.toString(buffer, sizeof(buffer)));
+               }
+            }
+         }
+      }
+
+   catch (const std::exception &e)
```

Comment:
I think there is a way to end-up with an entry in the _generatedClasses map that has a valid _namePrefix, but _classHashMap and _classPtrMap are both empty.
I saw in `invaliadateGeneratedClass` that, if _classHashMap and _classPtrMap become empty, we delete the corresponding key from _generatedClasses map.